jupyterhub/auth.py

```diff
@@ -141,6 +141,54 @@ def _deprecated_db(self):
         """
     ).tag(config=True)
 
+    allow_existing_users = Bool(
+        False,  # the default value is computed by a function below
+        config=True,
+        help="""
+        Allow existing users to login.
+
+        Defaults to True if `allowed_users` is set for historical reasons, and
+        False otherwise.
+
+        An existing user is a user in JupyterHub's database of users, and it
+        includes all users that has previously logged in.
+
+        .. warning::
+
+           Before enabling this you should review the existing users in the
+           JupyterHub admin panel at `/hub/admin`. You may find users existing
+           there because they have once been declared in config such as
+           `allowed_users`, or once been allowed to sign in.
+
+        .. warning::
+
+           When this is enabled and you are to remove access for one or more
+           users allowed via other config options, you must make sure that they
+           are not part of the database of users still. This can be tricky to do
+           if you stop allowing a group of externally managed users for example.
+
+        With this enabled, JupyterHub admin users can visit `/hub/admin` or use
+        JupyterHub's REST API to add and remove users as a way to allow or
+        revoke access.
+
+        The username for existing users must match the normalized username
+        returned by the authenticator.
+
+        .. versionadded:: 5.0
+        """,
+    )
+
+    @default("allow_existing_users")
+    def _allow_existing_users_default(self):
+        """
+        Computes the default value of allow_existing_users based on if
+        allowed_users to align with original behavior not introduce a breaking
+        change.
+        """
+        if self.allowed_users:
+            return True
+        return False
```

Comment:
Deal i'll expand this pr to include allow_all, giving it a dynamic default value that is true/false to not be breaking.

I'm not 100% confident its not breaking though if used with various authenticator logics, i'll have to think a bit more on if its non-breaking.

Reply:
Since we're preparing a major release, as long as it's not _usually_ breaking and desirable behavior is readily recoverable, and default behavior is at least as restrictive as now, I think we're okay.

I think that will be true, but thinking through the implications of the change with Authenticators is a good plan nonetheless, and should probably end up in the docs.

I think we should explicitly document that while `allow_all` is `not bool(allowed_users)` by default, Authenticator subclasses should:

1. be False by default for any authenticator that has a pool of users much larger than those who should have access (e.g. general user accounts, federated login, etc.),
2. need a custom dynamic default if they add any allow source beyond `allowed_users`

and include an example of a custom `@default("allow_all")`.

Reply:
I think probably the biggest, clearest change to make that simpler is to make allow_all only explicitly opt-in (I.e. effectively require all JupyterHub to have at least one allow config for anyone to be able to use it). I'm not a fan of that, but it does solve a hypothetical problem.

I'm less concerned about the scenario of removing the last user and then allowing unintentionally, because that doesn't seem very realistic, but it does make sense in principle. One way could be to make a clear distinction between _unset_ and specified but empty. I don't necessarily agree that it's surprising that going from specifying one user to not specifying allow_users at all would have this effect. But it _does_ seem reasonable to expect explicitly specifying an empty set to mean not allowing any users by name. We can investigate what that might look like, and what else might break as a result (e.g. use None as default).

Reply:
I'll also note that the defaults here are unchanged from current behavior. This PR aims to encode longstanding behavior in variables that can be overridden explicitly for clearer, predictable behavior in edge cases. So I think this PR ought to be purely an improvement in terms of simplicity and clarity without any changes to defaults.